src/sentry/ingest/billing_metrics_consumer.py

```diff
@@ -94,102 +94,173 @@ class MetricsBucket(TypedDict):
+
+            self._process_metrics_billing_bucket(time_left, force=True)
+
+        self._close_producers()
 
     def terminate(self) -> None:
         self.close()
 
+        if len(self._ongoing_billing_outcomes) > 0:
+            logger.warning(f"terminated, items dropped: {len(self._ongoing_billing_outcomes)}")
+        while self._ongoing_billing_outcomes:
+            ongoing_work, _ = self._ongoing_billing_outcomes.popleft()
+            if ongoing_work:
+                ongoing_work.cancel()
+
+        self._close_producers()
+
+    def _close_producers(self) -> None:
+        for producer in self._producers.values():
+            producer.close()
+
     def close(self) -> None:
-        self.__closed = True
+        self._closed = True
 
     def submit(self, message: Message[KafkaPayload]) -> None:
-        assert not self.__closed
-        self.__messages_since_last_commit += 1
+        assert not self._closed
+
+        if len(self._ongoing_billing_outcomes) >= self._max_buffer_size:
+            raise MessageRejected
 
-        payload = self._get_payload(message)
-        self._produce_billing_outcomes(payload)
-        self._mark_commit_ready(message)
+        bucket_payload = self._get_bucket_payload(message)
+        org_id = bucket_payload["org_id"]
+        quantity = self._count_processed_transactions(org_id, bucket_payload)
 
-    def _get_payload(self, message: Message[KafkaPayload]) -> MetricsBucket:
+        if quantity < 1:
+            # We still want to commmit buckets that don't generate billing
+            # outcomes, since the consumer has already processed them. To keep
+            # the offset order when committing, we still need to go through the
+            # futures queue.
+            future = None
+        else:
+            future = track_outcome_custom_publish(
+                self._produce,
+                org_id=org_id,
+                project_id=bucket_payload["project_id"],
+                key_id=None,
+                outcome=Outcome.ACCEPTED,
+                reason=None,
+                timestamp=datetime.now(),
+                event_id=None,
+                category=DataCategory.TRANSACTION,
+                quantity=quantity,
+            )
+
+        self._ongoing_billing_outcomes.append(MetricsBucketBilling(future, message))
+
+    def _get_bucket_payload(self, message: Message[KafkaPayload]) -> MetricsBucket:
         payload = json.loads(message.payload.value.decode("utf-8"), use_rapid_json=True)
         return cast(MetricsBucket, payload)
 
-    def _count_processed_transactions(self, bucket_payload: MetricsBucket) -> int:
-        if bucket_payload["metric_id"] != self.metric_id:
+    def _count_processed_transactions(self, org_id: int, bucket_payload: MetricsBucket) -> int:
+        counter_metric_id = indexer.resolve(
+            UseCaseKey.PERFORMANCE, org_id, self.counter_metric_name
+        )
+        if bucket_payload["metric_id"] != counter_metric_id:
```

Comment:
Another alternative here is to reverse resolve the metric id from the bucket and compare strings. I don't know what the performance (or other) consequences are for each of them, from Snuba to here.

Reply:
I think this could be a safer option actually. Extracting the mapping from the dict and comparing the string is negligible in terms of impact compared to the json parsing. So this will be a safer option than looking up the indexer at each message.

Reply:
@fpacifici do you mean reading the mapping from the [`mapping_meta`](https://github.com/getsentry/sentry/blob/adf4b85d50d7e0b3989c265abf2fce4b0e0e8a54/src/sentry/sentry_metrics/consumers/indexer/batch.py#L338) key in the bucket payload?